src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1080,10 +1081,35 @@ RouterInsertJob(Query *originalQuery, Query *query, DeferredErrorMessage **plann
 	Job *job = NULL;
 	bool requiresMasterEvaluation = false;
 	bool deferredPruning = false;
+	bool isMultiRowInsert = false;
 
-	if (!CanShardPrune(distributedTableId, query))
+	RangeTblEntry *valuesRTE = ExtractDistributedInsertValuesRTE(originalQuery);
+	if (valuesRTE != NULL)
 	{
-		/* there is a non-constant in the partition column, cannot prune yet */
+		/*
+		 * We expand the values_lists to contain all default expressions
+		 * from the target list. By doing this early on, in the original
+		 * query, we can later evaluate default expressions for each
+		 * individual row and then perform shard pruning.
+		 */
+		valuesRTE->values_lists = ExpandValuesLists(originalQuery->targetList,
+													valuesRTE->values_lists);
+
+		isMultiRowInsert = true;
+	}
+
+	if (isMultiRowInsert || !CanShardPrune(distributedTableId, query))
+	{
+		/*
+		 * If there is a non-constant (e.g. parameter, function call) in the partition
+		 * column of the INSERT then we defer shard pruning until the executor where
+		 * these values are known.
+		 *
+		 * XXX: We also defer pruning for multi-row INSERTs because of some current
```

Comment:
I think this is a pretty big caveat worth mentioning up front in the PR. The tradeoff here is that now _no_ multi-row inserts are prunable in the planning phase, even if they contain only constants. That seems like a tradeoff worth calling out.

Reply:
(I mean, there _is_ an `XXX` here, but I don't know the semantics of that other than "read this").

Reply:
We didn't do shard pruning in the planning phase before this PR either since CanShardPrune returns false if there is a non-Const (e.g. false) in the target entry of the partition column.

Reply:
The multi-row INSERT PR [changed that](https://github.com/citusdata/citus/blob/6a35c2937cd06f2604d9c59d701f520788ade9fb/src/backend/distributed/planner/multi_router_planner.c#L1167-L1177)… It uses the more abstract `ExtractInsertValuesList` to get a list of all values (whether they be sourced from `values_lists` or the target list) and iterates until it finds a non-`Const`.

Reply:
Ah true, but this didn't work correctly when there is an expression in a non-partition column, since we wouldn't evaluate it on the master. The query may still work because the worker can run functions like `now()`, though not sequences.